File: DotNetOnion/TorCryptoState.cs
using System;
using DotNetOnion.Crypto;
using DotNetOnion.Crypto.KDF;

namespace DotNetOnion
{
    internal class TorCryptoState
    {
        public TorStreamCipher forwardCipher { get; }
        public TorStreamCipher backwardCipher { get; }
        public TorMessageDigest forwardDigest { get; }
        public TorMessageDigest backwardDigest { get; }

        private TorCryptoState(TorStreamCipher forwardCipher, TorStreamCipher backwardCipher, TorMessageDigest forwardDigest, TorMessageDigest backwardDigest)
        {
            this.forwardCipher = forwardCipher;
            this.backwardCipher = backwardCipher;
            this.forwardDigest = forwardDigest;
            this.backwardDigest = backwardDigest;
        }
        internal static TorCryptoState CreateFromKdfResult(TorKdfResult kdfResult)
        {
            TorStreamCipher fCipher = new(kdfResult.ForwardKey);
            TorStreamCipher bCipher = new(kdfResult.BackwardKey);

            TorMessageDigest fDigest = new();
            TorMessageDigest bDigest = new();

            fDigest.Update(kdfResult.ForwardDigest);
            bDigest.Update(kdfResult.BackwardDigest);

            return new TorCryptoState(fCipher, bCipher, fDigest, bDigest);
        }
    }
}

File: NOnion/Crypto/TorCryptoState.fs
namespace NOnion.Crypto

open System

open NOnion.Crypto.Kdf

type TorCryptoState =
    {
        ForwardCipher: TorStreamCipher
        BackwardCipher: TorStreamCipher
        ForwardDigest: TorMessageDigest
        BackwardDigest: TorMessageDigest
        KeyHandshake: array<byte>
    }

    static member FromKdfResult
        (kdfResult: KdfResult)
        (reverse: bool)
        : TorCryptoState =
        let fCipher = TorStreamCipher(kdfResult.ForwardKey, None)
        let bCipher = TorStreamCipher(kdfResult.BackwardKey, None)
        let fDigest = TorMessageDigest(kdfResult.IsHSV3)
        let bDigest = TorMessageDigest(kdfResult.IsHSV3)

        fDigest.Update kdfResult.ForwardDigest 0 kdfResult.ForwardDigest.Length

        bDigest.Update
            kdfResult.BackwardDigest
            0
            kdfResult.BackwardDigest.Length

        if reverse then
            {
                ForwardCipher = bCipher
                BackwardCipher = fCipher
                ForwardDigest = bDigest
                BackwardDigest = fDigest
                KeyHandshake = kdfResult.KeyHandshake
            }
        else
            {
                ForwardCipher = fCipher
                BackwardCipher = bCipher
                ForwardDigest = fDigest
                BackwardDigest = bDigest
                KeyHandshake = kdfResult.KeyHandshake
            }
